x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrLicenseChecker.java

```diff
@@ -168,6 +187,41 @@ public void onFailure(final Exception e) {
                 });
     }
 
+    /**
+     * Fetches the history UUIDs for leader index on per shard basis using the specified leaderClient.
+     *
+     * @param leaderClient                              the leader client
+     * @param leaderIndex                               the name of the leader index
+     * @param onFailure                                 the failure consumer
+     * @param historyUUIDConsumer                       the leader index history uuid and consumer
+     * @param <T>                                       the type of response the listener is waiting for
+     */
+    // NOTE: Placed this method here; in order to avoid duplication of logic for fetching history UUIDs
+    // in case of following a local or a remote cluster.
+    public <T> void fetchLeaderHistoryUUIDs(
+        final Client leaderClient,
+        final String leaderIndex,
+        final Consumer<Exception> onFailure,
+        final Consumer<Map<Integer, String>> historyUUIDConsumer) {
+
+        CheckedConsumer<IndicesStatsResponse, Exception> indicesStatsHandler = indicesStatsResponse -> {
+            IndexStats indexStats = indicesStatsResponse.getIndices().get(leaderIndex);
+            Map<Integer, String> historyUUIDs = new HashMap<>();
```

Comment:
random drive by comment - we need to make sure we got history uuids for all the shards and a map data structure (compared to say, an array) hides it. Also - can we assert that the history uuids of all the shard copies that we got are identical?

Reply:
So then the index of the array indicates to which shard a historyUUID belongs?